### ISSUE TITLE
L2 normalization of a vector

### ISSUE THREAD

**Issue by rodrigob** [BVLC/caffe#1224]
Before implementing one more new layer from scratch, I want do double check.
I need to implement a vector normalization of the type `z / l2_norm(z)` it is there any way of doing this in current caffe-dev (or a related branch in the caffe network) ?

**Comment by seanbell** [comment]
@rodrigob I'm about to implement L2 normalization as well -- before I duplicate the effort, were you successful?

**Comment by sunbaigui** [comment]
@rodrigob @seanbell Looking forward to you guys' L2 norm layer, it'll be great if you guys could release it asap, time is money...

**Comment by kuprel** [comment]
Maybe try this?

https://github.com/kuprel/caffe/blob/master/src/caffe/layers/normalize_layer.cpp
https://github.com/kuprel/caffe/blob/master/src/caffe/layers/normalize_layer.cu

**Comment by sunbaigui** [comment]
@kuprel I've seen your code, it seems you've implemented angle loss for pair-wise learning. Does it go well, or could you share some experience on pair-wise learning?

**Comment by longjon** [comment]
Please ask usage questions on the [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) list. Thanks!

**Comment by hyojinie** [comment]
@seanbell Hi, is l2-normalization implemented?

**Comment by ducha-aiki** [comment]
@hyojinie  https://github.com/happynear/caffe-windows/blob/master/src/caffe/layers/normalize_layer.cpp

**Comment by jeffdonahue** [comment]
You can already do L2 normalization using something like this (untested but should work with minor syntax changes, I've used it before):

```
from caffe import layers as L, params as P

def l2normed(vec, dim):
    """Returns L2-normalized instances of vec; i.e., for each instance x in vec,
    computes  x / ((x ** 2).sum() ** 0.5). Assumes vec has shape N x dim."""
    denom = L.Reduction(vec, axis=1, operation=P.Reduction.SUMSQ)
    denom = L.Power(denom, power=(-0.5))
    denom = L.Reshape(denom, num_axes=0, axis=-1, shape=dict(dim=[1]))
    denom = L.Tile(denom, axis=1, tiles=dim)
    return L.Eltwise(vec, denom, operation=P.Eltwise.PROD)
```

**Comment by hyojinie** [comment]
@ducha-aiki Thanks. Have you used it before?
Would 
...
layer {
  name: "relu7"
  type: "ReLU"
  bottom: "fc7"
  top: "fc7"
}
layer {
  name: "nfeat"
  type: "NormalizeLayer"
  bottom: "fc7"
  top: "nfeat"
}

work?

**Comment by ducha-aiki** [comment]
@hyojinie, yes, it works. But you need to carefully initialize layers after the normalized, gaussian std=0.01 does not work :)

**Comment by hyojinie** [comment]
@ducha-aiki Thanks a lot. It works! I am trying to use it right before the loss layer (contrastive loss). I am guessing initialization wouldn't matter much in that case..?

**Comment by MenglaiWang** [comment]
@ducha-aiki  I have tried many gaussian std = 0.1,0.01.....,but it doesn't work. But uniform can do. Do you know why this happens?

**Comment by wangzeyu961130** [comment]
@hyojinie After adding "nfeat" layer and sending it to "fc8" classifier, the softmax loss decreases very slowly when training compared with no Normalize layer. The val accuracy first goes up and then down. Did anyone encounter this situation?

**Comment by sunsided** [comment]
@jeffdonahue How does the division work in that code? The `Reshape` + `Tile` magic is a bit too magic for me. :sweat_smile:

**Comment by xwang90** [comment]
@ducha-aiki Can I ask what you mean by "layers after the normalized"? Does it mean all the layers after the normalized layer?

**Comment by ducha-aiki** [comment]
@xwang90 yes

**Comment by xwang90** [comment]
@MenglaiWang Sorry to bother you. What do you mean by "gaussian std = 0.1,0.01.....,but it doesn't work"? "doesn't work" means low clasification accuracy? Thank you. :)

**Comment by xwang90** [comment]
Thanks a lot. So you mentioned "gaussian std=0.01 does not work", do you
mean that "gaussian std=0.01" would lead to low classification accuracy?
I just meet some low classification issues after incorporating this L2
normalization layer. Thanks again!

**Comment by soulslicer** [comment]
@hyojinie 

I too am trying to use it before contrastive loss. Did you manage to get this normalization layer to work?
